### PR TITLE
V7 send message API

### PR DESCRIPTION
**Pull request by EhsanKia**
defines a proper api for when a message is sent that modules can listen for and act on.

### CODE REVIEW

File: src/modules/emotes/emojis.js [hunk 1]
```diff
@@ -82,6 +83,15 @@ class Emojis extends AbstractEmotes {
                 this.emotes.set(code, emote);
             });
     }
+
+    convertEmojis(sendState) {
+        sendState.message =  sendState.message.split(' ').map(piece => {
```
Comment (night) [inline src/modules/emotes/emojis.js]: this is gonna create a chicken & the egg problem.. emojis run before something that replaces text that needs a replacement

File: src/modules/chat_commands/index.js [hunk 2]
```diff
@@ -234,46 +230,21 @@ function handleCommands(message) {
...
-        if (!emberView) return;
-
-        const newTwitchSendMessage = emberView._actions.sendMessage;
-
-        // check if we've already monkeypatched
-        if (newTwitchSendMessage === sendMessage) return;
+    handleSendMessage(sendState) {
+        const result = handleCommands(sendState.message);
+        if (result === false) {
+            sendState.message = '';
```
Comment (night) [inline src/modules/chat_commands/index.js]: this shouldn't be necessary.. to cancel sending a message why not a `sendState.cancel() which takes the clearing of the value in on its own

File: src/watcher.js [hunk 3]
```diff
@@ -195,7 +198,46 @@ class Watcher extends SafeEventEmitter {
             })
         );
 
+        let twitchSendMessage;
+
+        const emitSendMessage = sendState => this.emit('chat.send_message', sendState);
```
Comment (night) [inline src/watcher.js]: I honestly do not think this is the right solution to this problem. We need a separate module that functions as an ordered mutator for sending messages.

File: src/watcher.js [hunk 4]
```diff
@@ -195,7 +198,46 @@ class Watcher extends SafeEventEmitter {
...
+
+        function bttvSendMessage() {
+            try {
+                let defaultPrevented = false;
+                const sendState = {
```
Comment (night) [inline src/watcher.js]: this should be a message send class that gets instantiated on message send, don't ya think?

File: src/modules/emotes/emojis.js (hunk 1 above)
Comment (night) [inline src/modules/emotes/emojis.js]: also extra space

File: src/watcher.js [hunk 5]
```diff
@@ -13,6 +14,8 @@ const chatState = {
     r9k: 0,
     subsOnly: 0
 };
+const TEXTAREA_SELECTOR = '.textarea-contain';
+
```
Comment (night) [inline src/watcher.js]: extra new line

File: src/modules/chat_commands/index.js (hunk 2 above)
Comment (EhsanKia) [inline src/modules/chat_commands/index.js]: What if you want to abort the message send but not clear the textbox? Or are you proposing a third function that does both?

Comment (night) [inline src/modules/chat_commands/index.js]: what is the use case of an abort? i'm not opposed, just curious

Comment (EhsanKia) [inline src/modules/chat_commands/index.js]: Let's say anon chat blocks you from sending a message, we don't want to eat the user's message maybe? I guess once we add message history it's less of an issue, Also, maybe we want to transform the content of textbox for the user, but not send it yet.

Comment (night) [inline src/modules/chat_commands/index.js]: then i suggest having 2 is optimal

File: src/modules/tab_completion/index.js [hunk 6]
```diff
@@ -109,6 +116,27 @@ class TabCompletionModule {
         } else if (keyCode !== keyCodes.Shift) {
             this.tabTries = -1;
         }
+
+        // Message history
+        if (keyCode === keyCodes.UpArrow) {
```
Comment (night) [inline src/modules/tab_completion/index.js]: we need a way to disable this

Comment (EhsanKia) [inline src/modules/tab_completion/index.js]: Do we? It's pretty unobtrusive. Unlike the old one, it only goes to the previous if you're at the start of your message, and it also saves the edits you made the messages, so when you go back, you don't lose any data. The chances of it accidentally fucking you over are basically non-existent.

**Pull request merged**